c++/src/capnp/compat/http-over-capnp.c++

```diff
@@ -688,6 +834,69 @@ public:
     });
   }
 
+  kj::Promise<void> connect(ConnectContext context) override {
+    auto params = context.getParams();
+    auto host = params.getHost();
+    auto headers = factory.headersToKj(params.getHeaders());
+    auto pipe = kj::newTwoWayPipe();
+
+    class EofDetector final: public kj::AsyncOutputStream {
+    public:
+      EofDetector(kj::Own<kj::AsyncIoStream> inner) : inner(kj::mv(inner)) {}
+      ~EofDetector() {
+        inner->shutdownWrite();
+      }
+
+      kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(
+          kj::AsyncInputStream& input, uint64_t amount = kj::maxValue) override {
+        return inner->tryPumpFrom(input, amount);
+      }
+
+      kj::Promise<void> write(const void* buffer, size_t size) override {
+        return inner->write(buffer, size);
+      }
+
+      kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+        return inner->write(pieces);
+      }
+
+      kj::Promise<void> whenWriteDisconnected() override {
+        return inner->whenWriteDisconnected();
+      }
+    private:
+      kj::Own<kj::AsyncIoStream> inner;
+    };
+
+    kj::Own<kj::AsyncOutputStream> stream = factory.streamFactory.capnpToKj(
+        context.getParams().getDown());
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
+    // We write to the `down` pipe.
+    promises.add(pipe.ends[1]->pumpTo(*stream).then([context](uint64_t) mutable {
+      auto endReq = context.getParams().getDown().endRequest();
+      return endReq.send().ignoreResult();
+    }).attach(kj::mv(stream)).eagerlyEvaluate(
+        [&httpProxyStream = *pipe.ends[1]](kj::Exception&&) { httpProxyStream.abortRead(); }));
+
+    PipelineBuilder<ConnectResults> pb;
+    auto eofWrapper = kj::heap<EofDetector>(kj::mv(pipe.ends[1]));
```

Comment:
Since you're passing ownership of `pipe.ends[1]` to `EofDetector`, that means when `EofDetector` is destroyed, `pipe.ends[1]` will be destroyed. But the whole point of `EofDetector` is we expect it to be destroyed when the write direction is done, but the read direction could still be acti/e.

Unfortunately it seems like we need to add a refcounted wrapper around the pipe so that the pumps in both directions can hold references. The wrapper doesn't need to be a full `AsyncIoStream` implementation, it just needs to hold ownership. E.g.:

```
struct RefcountedWrapper: public kj::Refcounted {
  kj::Own<AsyncIoStream> stream;
};
auto& pipeRef = *pipe.ends[1];
auto refcounted = kj::refcounted<RefcountedWrapper>(kj::mv(pipe.ends[1]));
kj::Own<AsyncIoStream> ref1 = kj::attachRef(pipeRef, kj::addRef(*refcounted));
kj::Own<AsyncIoStream> ref2 = kj::attachRef(pipeRef, kj::addRef(*refcounted));
```

Now `ref1` and `ref2` are two refcounts on the same underlying stream.